### DISCUSSION TITLE
FBX Models from Blender don't draw in the correct order

### DISCUSSION THREAD

**Discussion by mghoffmann** [MonoGame/MonoGame#8399] [Graphics]
I made a simple test model in Blender and exported it as an FBX. The mesh has multiple parts, and the parts are not drawn in order. Here's a screenshot:

![image](https://github.com/MonoGame/MonoGame/assets/10456195/28147e5c-5264-4371-816e-f3bf16e9e6d7)

The bottom camera is positioned to the front and the right of the model, and the top camera is positioned directly behind it.

As you can see, the small sphere in front is drawing on top of the big sphere when the camera is behind it.

Do I need to do something to get depth buffering to apply to the parts? I'm just setting world, view, and projection matrices on the `BasicEffect` of each part and then calling `Model.Draw()`. Nothing fancy. Debugging shows that the loaded model only has one MeshPart.

Is this a problem with Blender's FBX exporter maybe? Microsoft Paint 3D renders the model as expected so I'm not sure.

**Comment by KelsamGames** [comment]
`GraphicsDevice.DepthStencilState = new DepthStencilState() { DepthBufferEnable = true };`

http://rbwhitaker.wikidot.com/2d3d-combined

**Reply by mghoffmann** [reply]
This helped, thanks.

I was already setting this in the start of the Game class's Draw method, but desk-checking it showed that it gets reset whenever `GraphicsDevice.Viewport` is changed. I'm using a different viewport for each camera so it was always set to `DepthStencilState.None`.

This is my fix:

```
GraphicsDevice.Viewport = camera.Viewport;
GraphicsDevice.DepthStencilState = DepthStencilState.Default;
```

**Reply by stromkos** [reply]
Be very careful with setting the `Viewport` value. Rendertargets, scissor and screensize changes may reset the values.

Each platform handles the  `Viewport` values differently. You may run into problems when porting your game.

**Reply by mghoffmann** [reply]
Good to know, thanks!
